dart/common/Composite.h

```diff
@@ -0,0 +1,212 @@
+  bool has() const;
+
+  /// Get a certain type of Aspect from this Composite
+  template <class T>
+  T* get();
+
+  /// Get a certain type of Aspect from this Composite
+  template <class T>
+  const T* get() const;
+
+  /// Make a clone of the aspect and place the clone into this Composite. If
+  /// an Aspect of the same type already exists in this Composite, the
+  /// existing Aspect will be destroyed.
+  template <class T>
+  void set(const T* aspect);
+
+  /// Use move semantics to place aspect into this Composite. If an Aspect of
+  /// the same type already exists in this Composite, the existing Aspect will
+  /// be destroyed.
+  template <class T>
+  void set(std::unique_ptr<T>&& aspect);
+
+  /// Construct an Aspect inside of this Composite
+  template <class T, typename ...Args>
+  T* createAspect(Args&&... args);
+
+  /// Remove an Aspect from this Composite.
+  template <class T>
+  void eraseAspect();
+
+  /// Remove an Aspect from this Composite, but return its unique_ptr instead
+  /// of letting it be deleted. This allows you to safely use move semantics to
+  /// transfer an Aspect between two Composites.
+  template <class T>
+  std::unique_ptr<T> releaseAspect();
+
+  /// Check if this Composite is specialized for a specific type of Aspect.
+  ///
+  /// By default, this simply returns false.
+  template <class T>
+  static constexpr bool isSpecializedFor();
+
+  /// Check if this Composite requires this specific type of Aspect
+  template <class T>
+  bool requiresAspect() const;
+
+  /// Set the states of the aspects in this Composite based on the given
+  /// Composite::State. The states of any Aspect types that do not exist
+  /// within this composite will be ignored.
+  void setCompositeState(const State& newStates);
+
+  /// Get the states of the aspects inside of this Composite
+  State getCompositeState() const;
+
+  /// Fill outgoingStates with the states of the aspects inside this Composite
+  void copyCompositeStateTo(State& outgoingStates) const;
+
+  /// Set the properties of the aspects in this Composite based on the given
+  /// Composite::Properties. The properties of any Aspect types that do not
+  /// exist within this composite will be ignored.
+  void setCompositeProperties(const Properties& newProperties);
+
+  /// Get the properties of the aspects inside of this Composite
+  Properties getCompositeProperties() const;
+
+  /// Fill outgoingProperties with the properties of the aspects inside this
+  /// Composite
+  void copyCompositePropertiesTo(Properties& outgoingProperties) const;
+
+  /// Give this Composite a copy of each Aspect from otherComposite
+  void duplicateAspects(const Composite* fromComposite);
+
+  /// Make the Aspects of this Composite match the Aspects of otherComposite. Any
+  /// Aspects in this Composite which do not exist in otherComposite will be
+  /// erased.
+  void matchAspects(const Composite* otherComposite);
+
+protected:
+
+  template <class T> struct type { };
```

Comment:
You're right. An older implementation of Aspect specialization required the base ``Composite`` class to use it, but that is no longer the case.